### PR TITLE
feat(codebuild): allow taking the artifact name from the buildspec

### LINKED ISSUE #5955: Expose the overrideArtifactName property on S3ArtifactsProps
Add the `overrideArtifactName` property to the `S3ArtifactsProps` interface.

### Use Case
The `overrideArtifactName` property enables dynamically naming artifacts from the project's buildspec.
- https://docs.aws.amazon.com/codebuild/latest/APIReference/API_ProjectArtifacts.html#CodeBuild-Type-ProjectArtifacts-overrideArtifactName
- https://docs.aws.amazon.com/codebuild/latest/userguide/build-spec-ref.html#build-spec-ref-syntax

### Proposed Solution
Include `overrideArtifactName` in the `S3ArtifactsProps` interface and include it in the compiled CloudFormation template.

### Other
N/A

* [x] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by shearn89**
### Commit Message
feat(codebuild): allow taking the artifact name from the buildspec

This commit adds support for CodeBuild Artifacts to have names
controlled by a buildspec file, allowing the use of shell scripting to
sensibly name artifacts in CodeBuild projects generated from the CDK.
The previosuly required field `name` in `S3ArtifactsProps` is now optional,
and not providing it will set the `overrideArtifactName` on the underlying L1 to `true`.

Fixes #5955
### End Commit Message

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by shearn89** [comment]
@skinny85 - Hi, any chance this can get reviewed in the next week or so? It's a feature that would unblock me at work! Hopefully it's a small enough change, despite the various merges from master.

**Comment by skinny85** [comment]
I'll review it today - apologies for the delay @shearn89 !

File: packages/@aws-cdk/aws-codebuild/lib/artifacts.ts [hunk 1]
```diff
@@ -121,6 +121,14 @@ export interface S3ArtifactsProps extends ArtifactsProps {
...
    */
   readonly encryption?: boolean;
+
+  /**
+   * If this is true, the name defined in the buildspec file will be used for artifacts.
+   * Otherwise, the artifact will use the <name> property.
+   *
+   * @default false - ignore buildspec artifact name
+   */
+  readonly useBuildspecName?: boolean;
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codebuild/lib/artifacts.ts:131]: I have 2 comments about this name:

1. We have a class called `BuildSpec`, so I think this should be called `useBuildSpecName`.
2. I wonder whether a name like `useNameFromBuildSpec`  or `useNameInBuildSpec` would be a little more descriptive?

File: packages/@aws-cdk/aws-codebuild/README.md [hunk 2]
```diff
@@ -273,6 +273,33 @@ const rule = project.onStateChange('BuildStateChange', {
...
+
+CodeBuild Projects can produce Artifacts and upload the to S3. For example:
+
+```ts
+const project = codebuild.Project(stack, 'MyProject', {
+  buildSpec: codebuild.BuildSpec.fromObject({
+    version: '0.2',
+  }),
+  artifacts:
+    codebuild.Artifacts.s3({
```
Comment (shearn89) [inline packages/@aws-cdk/aws-codebuild/README.md:286]: Looks good, think that's just a typo/failure of my IDE to indent properly!
```suggestion
  artifacts:  codebuild.Artifacts.s3({
```

[the remark this answers is not in the document]
Comment (shearn89) [reply packages/@aws-cdk/aws-codebuild/README.md:286]: Sounds good, I've commented again with the suggested code.

File: packages/@aws-cdk/aws-codebuild/lib/artifacts.ts (hunk 1 above)
Comment (shearn89) [reply packages/@aws-cdk/aws-codebuild/lib/artifacts.ts:131]: I like `useNameFromBuildSpec` - I'll update the code!

File: packages/@aws-cdk/aws-codebuild/README.md [hunk 3]
```diff
@@ -273,6 +273,33 @@ const rule = project.onStateChange('BuildStateChange', {
...
+      useBuildspecName: true,
+    }),
+});
+```
+
+This example will produce an artifact named as defined in the Buildspec file, 
+uploaded to an S3 bucket (`bucket`). The path will be `another/path` and the 
+artifact will be a zipfile. In this case, the `name` parameter will be ignored 
+by CodeBuild in favour of that specified in the `buildspec.yml`.
+
```
[the remark this answers is not in the document]
Comment (shearn89) [reply packages/@aws-cdk/aws-codebuild/README.md]: Done (I think)! See ed1c25b

File: packages/@aws-cdk/aws-codebuild/README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (shearn89) [reply packages/@aws-cdk/aws-codebuild/README.md]: Done! ed1c25b

File: packages/@aws-cdk/aws-codebuild/lib/artifacts.ts (hunk 1 above)
Comment (shearn89) [reply packages/@aws-cdk/aws-codebuild/lib/artifacts.ts]: Done! ed1c25b

**Comment by shearn89** [comment]
Hey @skinny85 - thanks for the review! I'm just about to push a commit with the proposed design change. I think it makes sense, and we shouldn't be changing any existing behaviour as the parameter is going from mandatory to optional. Given that this PR is the first time the `overrideArtifactName` parameter has been exposed I can't see that we'd be altering any existing behaviour!

File: packages/@aws-cdk/aws-codebuild/lib/artifacts.ts [hunk 4]
```diff
@@ -142,9 +149,10 @@ class S3Artifacts extends Artifacts {
         location: this.props.bucket.bucketName,
         path: this.props.path,
         namespaceType: this.props.includeBuildId === false ? 'NONE' : 'BUILD_ID',
-        name: this.props.name,
+        name: this.props.name == null ? project.node.uniqueId : this.props.name,
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codebuild/lib/artifacts.ts:152]: Did you test that this is necessary - we can't just leave this as `undefined`? `name` is optional in `CfnProject.ArtifactsProperty`: https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-codebuild.CfnProject.ArtifactsProperty.html#name

File: packages/@aws-cdk/aws-codebuild/README.md [hunk 5]
```diff
@@ -115,6 +115,30 @@ const bbSource = codebuild.Source.bitBucket({
 });
 ```
 
+## Artifacts
+
+CodeBuild Projects can produce Artifacts and upload the to S3. For example:
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codebuild/README.md:120]: Typo:

```suggestion
CodeBuild Projects can produce Artifacts and upload them to S3. For example:
```

File: packages/@aws-cdk/aws-codebuild/lib/artifacts.ts (hunk 4 above)
Comment (shearn89) [reply packages/@aws-cdk/aws-codebuild/lib/artifacts.ts:152]: I didn't - I'll do so today and set it to undefined if that works.

Comment (shearn89) [reply packages/@aws-cdk/aws-codebuild/lib/artifacts.ts]: Done, hopefully the integration tests pass!

Comment (skinny85) [reply packages/@aws-cdk/aws-codebuild/lib/artifacts.ts]: Just to make sure we're on the same page - by default, the integration tests act like snapshot tests. You need to specifically run the `npm integ` target to have them deployed. More info here: https://github.com/aws/aws-cdk/blob/master/CONTRIBUTING.md#integration-tests

**Comment by shearn89** [comment]
Ah, okay! I'll run those this evening and let you know the outcome...

File: packages/@aws-cdk/aws-codebuild/lib/artifacts.ts (hunk 4 above)
Comment (shearn89) [reply packages/@aws-cdk/aws-codebuild/lib/artifacts.ts]: Ah, okay! I'll run those this evening and let you know the outcome...

Comment (shearn89) [reply packages/@aws-cdk/aws-codebuild/lib/artifacts.ts]: Just ran the tests against my own account and all the `aws-codebuild` integration tests provision and tear down correctly, no errors visible in the CloudFormation console or build logs!

**Pull request merged**